<div style="text-align: center">
<img src="https://github.com/LinkedEarth/Logos/blob/master/PyleoTUPS/pyleotups_logo.png?raw=true" alt="PyleoTUPS logo" width="400">
</div>

# Querying data from the NOAA database

## Authors

Deborah Khider
<a href="https://orcid.org/0000-0001-7501-8430" target="_blank" rel="noopener noreferrer">
  <img src="https://orcid.org/sites/default/files/images/orcid_16x16.png" alt="ORCID iD" style="vertical-align: text-bottom;"/>
</a>

## Preamble

This tutorial will get you familiar with searching the NOAA database for relevant datasets and search them for relevant information such as variables, data, geographical information.

### Goals

 - Searching for a specific study and obtaining information such as location, publication, variable metadata and the associated data.
 - Searching over a geographical area

### Pre-requisites

### Reading time

Let's import our packages!

In [1]:
from pyleotups import Dataset
import pandas as pd

## From a study ID

This is the most basic search you can perform, using a NOAA ID to access a dataset. First, you need to create a `Dataset` object that will store the information.

In [2]:
ds=Dataset()

Let's do a simple search, knowing the NOAA study ID. For this example, let's use the dataset from [Clemens et al. (2021)](https://www.science.org/doi/10.1126/sciadv.abg3848), which can be accessed through [NOAA Paleo portal](https://www.ncei.noaa.gov/access/paleo-search/study/33213).

In [3]:
ds.search_studies(noaa_id=33213)

Parsing NOAA studies: 100%|██████████| 1/1 [00:00<00:00, 2785.06it/s]


StudyID  XMLID                                          StudyName  \
0   33213  74834  Bay of Bengal, Northeast Indian Margin Stable ...   

           DataType  EarliestYearBP  MostRecentYearBP  EarliestYearCE  \
0  PALEOCEANOGRAPHY         1462580               280        -1460630   

   MostRecentYearCE                                         StudyNotes  \
0              1670  Provided Keywords: Indian monsoon, South Asian...   

  ScienceKeywords                                      Investigators  \
0            None  Kaustubh Thirumalai, Liviu Giosan, Julie Riche...   

                                        Publications  \
0  [{'Author': 'Clemens, Steven; Yamamoto, Masano...   

                                               Sites  \
0  [[{'DataTableID': '45857', 'DataTableName': 'U...   

                                             Funding  
0  [{'fundingAgency': 'US National Science Founda...

The `summary` method provides basic information about the dataset, such as the name of the study, the NOAA [DataType](https://www.ncei.noaa.gov/products/paleoclimatology), the time coverage and the associated publication. The function retuns a `pandas.DataFrame`.

In [4]:
df = ds.get_summary()
df

StudyID  XMLID                                          StudyName  \
0   33213  74834  Bay of Bengal, Northeast Indian Margin Stable ...   

           DataType  EarliestYearBP  MostRecentYearBP  EarliestYearCE  \
0  PALEOCEANOGRAPHY         1462580               280        -1460630   

   MostRecentYearCE                                         StudyNotes  \
0              1670  Provided Keywords: Indian monsoon, South Asian...   

  ScienceKeywords                                      Investigators  \
0            None  Kaustubh Thirumalai, Liviu Giosan, Julie Riche...   

                                        Publications  \
0  [{'Author': 'Clemens, Steven; Yamamoto, Masano...   

                                               Sites  \
0  [[{'DataTableID': '45857', 'DataTableName': 'U...   

                                             Funding  
0  [{'fundingAgency': 'US National Science Founda...

`PyleoTUPS` allows you to get metadata information that will be returned in separate dataframes. For instance, let's look at the information about the publication associated with the dataset. This function also returns BibTeX entries along with the DataFrame.

In [8]:
bib, df = ds.get_publications()
df

Author  \
0  Clemens, Steven; Yamamoto, Masanobu; Thirumala...   

                                               Title           Journal  Year  \
0  Remote and Local Drivers of Pleistocene South ...  Science Advances  2021   

  Volume Number  Pages         Type                     DOI  \
0      7     23    NaN  publication  10.1126/sciadv.abg3848   

                                        URL           CitationKey StudyID  \
0  http://dx.doi.org/10.1126/sciadv.abg3848  M._Remote_2021_33213   33213   

                                           StudyName  
0  Bay of Bengal, Northeast Indian Margin Stable ...

<div style="border-left: 4px solid #1f77b4; background: #f0f8ff; padding: 0.5em 1em; border-radius: 4px;">
  <strong>Note:</strong> You can set the `save` parameter to True if you want to save a copy of the BibTeX entries.
</div>

Similarly, you can return information about the location of the record: 

In [5]:
df_geo = ds.get_geo()
df_geo

StudyID          DataType SiteID    SiteName        LocationName Latitude  \
0   33213  PALEOCEANOGRAPHY  58697  IODP U1446  Ocean>Indian Ocean   19.083   

  Longitude MinElevation MaxElevation  
0    85.733        -1440        -1440

And the funding that supported the study:

In [7]:
df_fund = ds.get_funding()
df_fund

StudyID                                          StudyName  \
0   33213  Bay of Bengal, Northeast Indian Margin Stable ...   
1   33213  Bay of Bengal, Northeast Indian Margin Stable ...   
2   33213  Bay of Bengal, Northeast Indian Margin Stable ...   
3   33213  Bay of Bengal, Northeast Indian Margin Stable ...   

                                       FundingAgency  \
0                     US National Science Foundation   
1  Japan Society for the Promotion of Science (JSPS)   
2     UK Natural Environment Research Council (NERC)   
3             United States Geological Survey (USGS)   

                FundingGrant  
0                 OCE1634774  
1            JPMXS05R2900001  
2  JPMXS05R2900001, 19H05595  
3               NE/L002493/1

<div style="border-left: 4px solid #1f77b4; background: #f0f8ff; padding: 0.5em 1em; border-radius: 4px;">
  <strong>Note:</strong> As you may have noticed, each table has the NOAA study ID, this is useful when you are looking at multiple studies after performing a search for multiple datasets.
</div>


Next, let's have a look at the tables of data present in the dataset:

In [9]:
df_tables = ds.get_tables()
df_tables

DataTableID                            DataTableName   TimeUnit  \
0       45857       U1446 Benthic Isotopes Clemens2021  cal yr BP   
1       45858      U1446 Planktic Isotopes Clemens2021  cal yr BP   
2       45859             U1446 TEX86H_SST Clemens2021  cal yr BP   
3       45860                 U1446 d18Osw Clemens2021  cal yr BP   
4       45861  U1446 LeafWax CarbonIsotope Clemens2021  cal yr BP   
5       45862                  U1446 Mg/Ca Clemens2021  cal yr BP   
6       45863                  U1446 Rb/Ca Clemens2021  cal yr BP   
7       45864              U1446 Age Model Clemens2021  cal yr BP   

                                             FileURL  \
0  https://www.ncei.noaa.gov/pub/data/paleo/contr...   
1  https://www.ncei.noaa.gov/pub/data/paleo/contr...   
2  https://www.ncei.noaa.gov/pub/data/paleo/contr...   
3  https://www.ncei.noaa.gov/pub/data/paleo/contr...   
4  https://www.ncei.noaa.gov/pub/data/paleo/contr...   
5  https://www.ncei.noaa.gov/pub/data/paleo/contr...   
6  https://www.ncei.noaa.gov/pub/data/paleo/contr...   
7  https://www.ncei.noaa.gov/pub/data/paleo/contr...   

                                           Variables     FileDescription  \
0  [Site, Hole, Type, Section, Core, Section_Dept...  NOAA Template File   
1  [Site, Hole, Type, Section, Comment, Core, Sec...  NOAA Template File   
2  [Site, Hole, Type, Section, Core, Section_Dept...  NOAA Template File   
3  [Age, SL_Scaled, SL_Scaled_averaged, d18O_G_ru...  NOAA Template File   
4  [Site, Hole, Type, Section, Core, Section_Dept...  NOAA Template File   
5  [Site, Hole, Type, Analytical_Facility, Core, ...  NOAA Template File   
6  [Site, Hole, Type, Section, Core, Section_Dept...  NOAA Template File   
7                [Site, Comments, Sample_Depth, Age]  NOAA Template File   

   TotalFilesAvailable SiteID    SiteName        LocationName Latitude  \
0                    1  58697  IODP U1446  Ocean>Indian Ocean   19.083   
1                    1  58697  IODP U1446  Ocean>Indian Ocean   19.083   
2                    1  58697  IODP U1446  Ocean>Indian Ocean   19.083   
3                    1  58697  IODP U1446  Ocean>Indian Ocean   19.083   
4                    1  58697  IODP U1446  Ocean>Indian Ocean   19.083   
5                    1  58697  IODP U1446  Ocean>Indian Ocean   19.083   
6                    1  58697  IODP U1446  Ocean>Indian Ocean   19.083   
7                    1  58697  IODP U1446  Ocean>Indian Ocean   19.083   

  Longitude MinElevation MaxElevation StudyID  \
0    85.733        -1440        -1440   33213   
1    85.733        -1440        -1440   33213   
2    85.733        -1440        -1440   33213   
3    85.733        -1440        -1440   33213   
4    85.733        -1440        -1440   33213   
5    85.733        -1440        -1440   33213   
6    85.733        -1440        -1440   33213   
7    85.733        -1440        -1440   33213   

                                           StudyName  
0  Bay of Bengal, Northeast Indian Margin Stable ...  
1  Bay of Bengal, Northeast Indian Margin Stable ...  
2  Bay of Bengal, Northeast Indian Margin Stable ...  
3  Bay of Bengal, Northeast Indian Margin Stable ...  
4  Bay of Bengal, Northeast Indian Margin Stable ...  
5  Bay of Bengal, Northeast Indian Margin Stable ...  
6  Bay of Bengal, Northeast Indian Margin Stable ...  
7  Bay of Bengal, Northeast Indian Margin Stable ...

The dataset contains 8 tables. To get the data, you need to pass the `DataTableID` or `FileURL` to the `get_data()` method. Let's have a look at the TEX86 data:

In [10]:
dfs = ds.get_data(dataTableIDs="45859")
dfs[0].head()

Site Hole Core Type Section Section_Depth Sample_Depth   Age   TEX86H  \
0  U1446    C    1    H       1           4.5        0.045  0.31  -0.1177   
1  U1446    C    1    H       1          31.5        0.315  0.66  -0.1216   
2  U1446    C    1    H       1          61.5        0.615  1.04  -0.1183   
3  U1446    C    1    H       1          90.5        0.905  1.41  -0.1089   
4  U1446    C    1    H       1         121.5        1.215  1.80  -0.1155   

     SST  
0  30.55  
1  30.28  
2  30.51  
3  31.15  
4  30.70

<div style="border-left: 4px solid #1f77b4; background: #f0f8ff; padding: 0.5em 1em; border-radius: 4px;">
  <strong>Note:</strong> You can pass multiple tables IDs as a list. The function will always return a list of DataFrames (hence why we selected the first one in the code above.)
</div>

Some relevant metadata for each column is stored in the DataFrame attributes:

In [12]:
df.attrs

{'variables': ['Site',
  'Hole',
  'Core',
  'Type',
  'Section',
  'Section_Depth',
  'Sample_Depth',
  'Age',
  'TEX86H',
  'SST'],
 'NOAAStudyId': '33213',
 'StudyName': 'Bay of Bengal, Northeast Indian Margin Stable Isotope, Biomarker and SST Reconstructions since the Mid-Pleistocene'}

To get more information about the variables, you can use the `get_variables` function:

In [13]:
df_var = ds.get_variables(dataTableIDs="45859")
df_var

StudyID SiteID                                            FileURL  \
DataTableID                                                                     
45859         33213  58697  https://www.ncei.noaa.gov/pub/data/paleo/contr...   
45859         33213  58697  https://www.ncei.noaa.gov/pub/data/paleo/contr...   
45859         33213  58697  https://www.ncei.noaa.gov/pub/data/paleo/contr...   
45859         33213  58697  https://www.ncei.noaa.gov/pub/data/paleo/contr...   
45859         33213  58697  https://www.ncei.noaa.gov/pub/data/paleo/contr...   
45859         33213  58697  https://www.ncei.noaa.gov/pub/data/paleo/contr...   
45859         33213  58697  https://www.ncei.noaa.gov/pub/data/paleo/contr...   
45859         33213  58697  https://www.ncei.noaa.gov/pub/data/paleo/contr...   
45859         33213  58697  https://www.ncei.noaa.gov/pub/data/paleo/contr...   
45859         33213  58697  https://www.ncei.noaa.gov/pub/data/paleo/contr...   

              VariableName                                cvDataType  \
DataTableID                                                            
45859                 Site                          PALEOCEANOGRAPHY   
45859                 Hole                          PALEOCEANOGRAPHY   
45859                 Type                          PALEOCEANOGRAPHY   
45859              Section                          PALEOCEANOGRAPHY   
45859                 Core                          PALEOCEANOGRAPHY   
45859        Section_Depth                          PALEOCEANOGRAPHY   
45859         Sample_Depth  CLIMATE RECONSTRUCTIONS|PALEOCEANOGRAPHY   
45859                  Age  CLIMATE RECONSTRUCTIONS|PALEOCEANOGRAPHY   
45859               TEX86H                          PALEOCEANOGRAPHY   
45859                  SST  CLIMATE RECONSTRUCTIONS|PALEOCEANOGRAPHY   

                                                        cvWhat  \
DataTableID                                                      
45859                  sampling metadata>sample identification   
45859                  sampling metadata>sample identification   
45859                  sampling metadata>sample identification   
45859                  sampling metadata>sample identification   
45859                  sampling metadata>sample identification   
45859                  sampling metadata>sample identification   
45859                                     depth variable>depth   
45859                                         age variable>age   
45859        chemical composition>compound>organic compound...   
45859        earth system variable>temperature variable>tem...   

                                                    cvMaterial cvError  \
DataTableID                                                              
45859                                                     None    None   
45859                                                     None    None   
45859                                                     None    None   
45859                                                     None    None   
45859                                                     None    None   
45859                                                     None    None   
45859                                                     None    None   
45859                                                     None    None   
45859                                                     None    None   
45859        reconstruction material>organic compound index...    None   

                                                        cvUnit cvSeasonality  \
DataTableID                                                                    
45859                                                     None          None   
45859                                                     None          None   
45859                                                     None          None   
45859                                                     None          None   
45859       

Instead of the table ID, you can also pass the file URL to get the data:

In [14]:
df1 = ds.get_data(file_urls="https://www.ncei.noaa.gov/pub/data/paleo/contributions_by_author/clemens2021/clemens2021-u1446-mgca-noaa.txt")[0]
df1.head()

Site Hole Core Type Section Section_Depth Sample_Depth   Age Mg/Ca  \
0  U1446    C    1    H       1             5         0.05  0.32  4.63   
1  U1446    C    1    H       1            32         0.32  0.66  4.62   
2  U1446    C    1    H       1            62         0.62  1.05  4.72   
3  U1446    C    1    H       1            91         0.91  1.42  4.58   
4  U1446    C    1    H       1           122         1.22  1.81  4.70   

   Analytical_Facility Mg/Ca_SST  
0  Rosenthal (Rutgers)     28.41  
1  Rosenthal (Rutgers)     28.46  
2  Rosenthal (Rutgers)     28.43  
3  Rosenthal (Rutgers)     28.63  
4  Rosenthal (Rutgers)     28.61

This shows you how you can retrieve information about a dataset from NOAA and get information about the dataset. However, in most cases, you may not know the NOAA ID. NOAA offers an API search across several query parameters such as geographical extent, datatype... 

Let's have a look!

## Geographical Query

Let's start with a simple geographical query. Let's look for all the datasets within 5°S-5°N and 109-125°E, roughly corresponding to the Indo-Pacific Warm Pool. 

Let's create another `Dataset` object to store this new search:

In [17]:
ds2 = Dataset()

Querying terms and their definitions are available through NOAA: https://www.ncei.noaa.gov/access/paleo-search/api

In [16]:
ds2.search_studies(max_lat=5, min_lat=-5, max_lon=109,
                       min_lon=125)

Parsing NOAA studies: 100%|██████████| 100/100 [00:00<00:00, 21706.28it/s]


StudyID  XMLID                                          StudyName  \
0    11194   9632  1,100 Year El Niño/Southern Oscillation (ENSO)...   
1    22031  20009  1200 Year Atlantic Multidecadal Variability an...   
2    39047  80187  1500 Year Sedimentological and Geochemical Dat...   
3     2614   1685  350 KYr Sea Level Reconstruction and Foraminif...   
4    14632  12613  700 Year El Niño/Southern Oscillation (ENSO) N...   
..     ...    ...                                                ...   
95   25611  23814  Eastern Equatorial Pacific d18O and d13C Data ...   
96   30152  71852  Eastern Pacific Ocean TEX86 and Mg/Ca SST Reco...   
97    2624   1695  Eastern Pacific Pleistocene Alkenone Data and ...   
98   19139  16805  Eastern Tropical Indian Ocean 45,000 Year d18O...   
99   19874  17739  Eastern Tropical Pacific 1,000 Year Foraminfer...   

                   DataType  EarliestYearBP  MostRecentYearBP  EarliestYearCE  \
0   CLIMATE RECONSTRUCTIONS          1050.0             -52.0           900.0   
1   CLIMATE RECONSTRUCTIONS          1150.0             -60.0           800.0   
2            PALEOLIMNOLOGY          1200.0             -50.0           750.0   
3          PALEOCEANOGRAPHY        361500.0            1000.0       -359550.0   
4   CLIMATE RECONSTRUCTIONS           649.0             -55.0          1301.0   
..                      ...             ...               ...             ...   
95         PALEOCEANOGRAPHY         24734.0               0.0        -22784.0   
96  CLIMATE RECONSTRUCTIONS         24734.0            1239.0        -22784.0   
97         PALEOCEANOGRAPHY       1833700.0            4472.0      -1831750.0   
98         PALEOCEANOGRAPHY         45330.0              60.0        -43380.0   
99         PALEOCEANOGRAPHY          1088.0             -59.0           862.0   

    MostRecentYearCE                                         StudyNotes  \
0             2002.0  An index of canonical ENSO variability for the...   
1             2010.0  Summer (May-September) Atlantic Multidecadal V...   
2             2000.0  Elevations of lakes: Tota = 3015m; Siscunsi = ...   
3              950.0                                                      
4             2005.0  An index of canonical El Niño/Southern Oscilla...   
..               ...                                                ...   
95            1950.0                                               None   
96             711.0  Metadata are from the Temperature-12k project ...   
97           -2522.0                                                      
98            1890.0  High-resolution (~30-80 years) foraminiferal o...   
99            2009.0  Oxygen isotopes and Mg/Ca data from planktonic...   

                                      ScienceKeywords  \
0   [Atmospheric and Oceanic Circulation Patterns ...   
1   [Atmospheric and Oceanic Circulation Patterns ...   
2                      [Precipitation Reconstruction]   
3                          [Sea Level Reconstruction]   
4   [ENSO, Atmospheric and Oceanic Circulation Pat...   
..                                                ...   
95                             [Last Glacial Maximum]   
96                                               None   
97           [Sea Surface Temperature Reconstruction]   
98           [Sea Surface Temperature Reconstruction]   
99       [Little Ice Age (LIA), Medieval Warm Period]   

                                        Investigators  \
0   Jinbao Li, Shang-Ping Xie, Edward Cook, Rosann...   
1   Jianglin Wang, Bao Yang, Fredrik Ljungqvist, J...   
2   Broxton Bird, Byron Steinman, Jaime Escobar, A...   
3   David Lea, Pamela Martin, Dorothy Pak, Howard ...   
4   Jinbao Li, Shang-Ping Xie, Edward Cook, Marian...   
..                                                ...   
95  Heather Ford, Celia McChesney, Jennifer Hertzb...   
96  Jennifer Hertzberg, Matthew Schmidt, Thomas Bi...   
97                      Zhonghui Liu, Timothy Herbert   
98  Mahya

<div style="border-left: 4px solid #d62728; background: #ffe6e6; padding: 0.5em 1em; border-radius: 4px;">
  <strong>Warning:</strong> One of the default parameters in `search_studies` is how many studies can be returned by query. The default is 100. 
</div>

We got 100 datasets, which means we may have hit the limit. Let's increase to 500 and see what happens:

In [18]:
ds2.search_studies(max_lat=5, min_lat=-5, max_lon=109,
                       min_lon=125, limit=500)

Parsing NOAA studies: 100%|██████████| 383/383 [00:00<00:00, 7918.97it/s]


StudyID  XMLID                                          StudyName  \
0     11194   9632  1,100 Year El Niño/Southern Oscillation (ENSO)...   
1     22031  20009  1200 Year Atlantic Multidecadal Variability an...   
2     39047  80187  1500 Year Sedimentological and Geochemical Dat...   
3      2614   1685  350 KYr Sea Level Reconstruction and Foraminif...   
4     14632  12613  700 Year El Niño/Southern Oscillation (ENSO) N...   
..      ...    ...                                                ...   
378    6400   2421  Western Tropical Pacific, 21Kyr BP to present,...   
379   16791  14498  Western Ugandan Crater Lakes 1000Yr Diatom-inf...   
380    8722   2801  Yasuhara et al. 2009 Equatorial Atlantic ODP92...   
381    6266   2549  de Garidel-Thoron et al. 2005 Western Pacific ...   
382    8678   2764  de Garidel-Thoron et al. 2007 Warm Pool 30ka M...   

                    DataType  EarliestYearBP  MostRecentYearBP  \
0    CLIMATE RECONSTRUCTIONS          1050.0             -52.0   
1    CLIMATE RECONSTRUCTIONS          1150.0             -60.0   
2             PALEOLIMNOLOGY          1200.0             -50.0   
3           PALEOCEANOGRAPHY        361500.0            1000.0   
4    CLIMATE RECONSTRUCTIONS           649.0             -55.0   
..                       ...             ...               ...   
378         PALEOCEANOGRAPHY         21921.0               2.0   
379           PALEOLIMNOLOGY           878.0             -57.0   
380         PALEOCEANOGRAPHY        500000.0               0.0   
381  CLIMATE RECONSTRUCTIONS       1750000.0               0.0   
382         PALEOCEANOGRAPHY         30000.0               0.0   

     EarliestYearCE  MostRecentYearCE  \
0             900.0            2002.0   
1             800.0            2010.0   
2             750.0            2000.0   
3         -359550.0             950.0   
4            1301.0            2005.0   
..              ...               ...   
378        -19971.0            1948.0   
379          1072.0            2007.0   
380       -498050.0            1950.0   
381      -1748050.0            1950.0   
382        -28050.0            1950.0   

                                            StudyNotes  \
0    An index of canonical ENSO variability for the...   
1    Summer (May-September) Atlantic Multidecadal V...   
2    Elevations of lakes: Tota = 3015m; Siscunsi = ...   
3                                                        
4    An index of canonical El Niño/Southern Oscilla...   
..                                                 ...   
378  Foraminiferal stable isotope and Mg/Ca data fo...   
379  Diatom census data and reconstructed Lake Leve...   
380                                               None   
381                                               None   
382                                               None   

                                       ScienceKeywords  \
0    [Atmospheric and Oceanic Circulation Patterns ...   
1    [Atmospheric and Oceanic Circulation Patterns ...   
2                       [Precipitation Reconstruction]   
3                           [Sea Level Reconstruction]   
4    [ENSO, Atmospheric and Oceanic Circulation Pat...   
..                                                 ...   
378                                               None   
379                                        [hydrology]   
380                                               None   
381           [Sea Surface Temperature Reconstruction]   
382           [Sea Surface Temperature Reconstruction]   

                                         Investigators  \
0    Jinbao Li, Shang-Ping Xie, Edward Cook, Rosann...   
1    Jianglin Wang, Bao Yang, Fredrik Ljungqvist, J...   
2    Broxton Bird, Byron Steinman, Jaime Escobar, A...   
3    David Lea, Pamela Martin, Dorothy Pak, Howard ...   
4    Jinbao Li, Shang-Ping Xie, Edward Cook, Marian...   
..                                                 ...   
378      Axel Timmermann, Lowell Stott, R

We have 383 studies now, far from the maximum of 500 so it seams we collected all records. 

You can use the same functions as described above the retrieve information. For instance, let's have a look at the data tables:

In [19]:
df_tables_geo = ds2.get_tables()
df_tables_geo.head()

DataTableID            DataTableName                      TimeUnit  \
0       19791              enso_li2011                            CE   
1       33108          Wang2017AMV-AMO                            CE   
2       51864           MCEOF Bird2024                     cal yr BP   
3       55412     bird2024 RecordsUsed  calendar year before present   
4       51863  Tota C/N, Clay Bird2024                     cal yr BP   

                                             FileURL  \
0  https://www.ncei.noaa.gov/pub/data/paleo/treer...   
1  https://www.ncei.noaa.gov/pub/data/paleo/recon...   
2  https://www.ncei.noaa.gov/pub/data/paleo/paleo...   
3  https://www.ncei.noaa.gov/pub/data/paleo/paleo...   
4  https://www.ncei.noaa.gov/pub/data/paleo/paleo...   

                                           Variables  \
0                              [age, ensoi, ensovar]   
1  [age, Atlantic Multidecadal Variability Index,...   
2  [pumacocha_1sig_hi, pumacocha_2sig_hi, quelcca...   
3                                                 []   
4      [Tota_avg_depth, Tota_H13_%clay, Tota_H13_cn]   

            FileDescription  TotalFilesAvailable SiteID  \
0        NOAA Template File                    1  48144   
1  Formatted Text Data File                    1  56395   
2        NOAA Template File                    1  60174   
3        NOAA Template File                    1  60174   
4        NOAA Template File                    1  60952   

                 SiteName                               LocationName  \
0        Tropical Pacific                        Ocean>Pacific Ocean   
1    North Atlantic Ocean  Ocean>Atlantic Ocean>North Atlantic Ocean   
2  Circum-Andes Synthesis                    Continent>South America   
3  Circum-Andes Synthesis                    Continent>South America   
4               Lake Tota           Continent>South America>Colombia   

   Latitude   Longitude MinElevation MaxElevation StudyID  \
0       -20          20         None         None   11194   
1         0          70         None         None   22031   
2   -13.933      10.712         3015         3680   39047   
3   -13.933      10.712         3015         3680   39047   
4  5.497382  -72.978709         3015         3015   39047   

                                           StudyName  
0  1,100 Year El Niño/Southern Oscillation (ENSO)...  
1  1200 Year Atlantic Multidecadal Variability an...  
2  1500 Year Sedimentological and Geochemical Dat...  
3  1500 Year Sedimentological and Geochemical Dat...  
4  1500 Year Sedimentological and Geochemical Dat...

Other searches possible under the `search_studies` method:

In [20]:
?ds.search_studies

Signature:
ds.search_studies(
    xml_id=None,
    noaa_id=None,
    data_publisher='NOAA',
    data_type_id=None,
    keywords=None,
    investigators=None,
    max_lat=None,
    min_lat=None,
    max_lon=None,
    min_lon=None,
    location=None,
    publication=None,
    search_text=None,
    earliest_year=None,
    latest_year=None,
    cv_whats=None,
    recent=False,
    limit=100,
)
Docstring:
Search for NOAA studies using the specified parameters.

At least one parameter must be provided to perform a search. This method interfaces with
the NOAA NCEI Paleo Study Search API. Use it to filter studies based on location,
investigators, time range, keywords, and more.

Parameters
----------
xml_id : str, optional
    Specify the internal XML document ID. Must be an exact match (e.g., '1840').

noaa_id : str, optional
    Provide the unique NOAA Study ID as a number (e.g., '13156').

search_text : str, optional
    General text search across study content. Supports wildcards (%) and l